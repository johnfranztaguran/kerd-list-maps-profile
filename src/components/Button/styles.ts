import { StyleSheet, TextStyle, ViewStyle, ImageStyle } from 'react-native'
import { fonts, dimens, colors } from 'constant'

type Styles = {
  buttonLabel: TextStyle
  buttonBorder: TextStyle
  buttonIconContent: ViewStyle
  buttonIconSize: ImageStyle
  buttonIconLabel: ViewStyle
}

const styles: Styles = StyleSheet.create<Styles>({
  buttonLabel: {
    /* 14 sp â€¢ BUTTON */
    fontFamily: fonts.OpenSansRegular,
    fontStyle: 'normal',
    fontWeight: 'normal',
    fontSize: dimens.fontSize.size_14,
    lineHeight: 19,
    marginTop: 12,
    marginBottom: 12,
    display: 'flex',
    alignItems: 'center',
    textAlign: 'center',
    letterSpacing: 0.25,
    textTransform: 'capitalize'
  },
  buttonBorder: {
    borderRadius: 4
  },
  buttonIconContent: {
    paddingRight: 12,
    flexDirection: 'row',
    justifyContent: 'space-between',
    color: colors.pigeonPost
  },
  buttonIconSize: {
    width: 14,
    height: 14
  },
  buttonIconLabel: {
    marginStart: 14,
    marginEnd: 0
  }
})

export default styles
