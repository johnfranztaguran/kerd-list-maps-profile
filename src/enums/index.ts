export enum Screens {
  Main = 'Main',
  Login = 'Login',
  Register = 'Register',
  CompleteRegister = 'CompleteRegister',
  Welcome = 'Welcome',
  ListForm = 'ListForm',
  ProfileSettings = 'ProfileSettings',
  GoogleMaps = 'GoogleMaps',
  ListCrud = 'ListCrud',
  ProfileCrud = 'ProfileCrud'
}

export enum TextInputType {
  Number = 'number',
  Decimal = 'decimal',
  Text = 'text',
  Email = 'email',
  Password = 'password'
}

export enum MessageType {
  Error = 'error',
  Info = 'info'
}

export enum ErrorType {
  Empty = 'empty',
  Invalid = 'invalid',
  Existed = 'existed',
  Info = 'info',
  InvalidCombination = 'invalidCombination'
}

export enum HelperMessageType {
  Info = 'info',
  Error = 'error'
}

export enum ButtonMode {
  Outlined = 'outlined',
  Contained = 'contained',
  Text = 'text'
}

export enum ButtonType {
  PlainButton = 'button',
  ButtonIcon = 'button-icon'
}
