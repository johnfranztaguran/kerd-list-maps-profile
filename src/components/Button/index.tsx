import React from 'react'
import { Image } from 'react-native'
import { Button as RNPButton } from 'react-native-paper'
import { ImageSource } from 'react-native-vector-icons/Icon'
import { testIdProps } from 'utils/testAutomation'
import { ButtonMode, ButtonType } from 'enums'
import { strings, colors, themes } from 'constant'
import styles from './styles'

type Props = {
  mode?: ButtonMode
  type?: ButtonType
  text: string
  color?: string
  textColor?: string
  borderColor?: string
  onPress: () => void
  icon?: ImageSource
}

const Button = (props: Props): JSX.Element => {
  const {
    mode = ButtonMode.Contained,
    type = ButtonType.PlainButton,
    text,
    color = themes.colors.primary,
    textColor = colors.solidWhite,
    borderColor = colors.transparent,
    onPress,
    icon
  } = props

  const generateField = (): JSX.Element => {
    switch (type) {
      case ButtonType.PlainButton:
        return createPlainButton()
      case ButtonType.ButtonIcon:
        return createIconButton()
      default:
        return createPlainButton()
    }
  }

  const createPlainButton = (): JSX.Element => {
    return (
      <RNPButton
        {...testIdProps(strings.testIdText.prefixBtn + text)}
        uppercase={false}
        mode={mode}
        color={color}
        style={[styles.buttonBorder, { borderColor }]}
        labelStyle={[styles.buttonLabel, { color: textColor }]}
        onPress={onPress}>
        {text}
      </RNPButton>
    )
  }

  const createIconButton = (): JSX.Element => {
    return (
      <RNPButton
        {...testIdProps(strings.testIdText.prefixBtn + text)}
        mode={mode}
        dark
        compact
        uppercase={false}
        style={styles.buttonBorder}
        labelStyle={styles.buttonIconLabel}
        contentStyle={styles.buttonIconContent}
        onPress={onPress}
        // TODO: find a way to load svg file dinamically
        icon={({ color: tintColor }) => (
          <Image
            source={icon}
            style={[
              {
                tintColor
              },
              styles.buttonIconSize
            ]}
          />
        )}>
        {text}
      </RNPButton>
    )
  }

  return generateField()
}

export default Button
