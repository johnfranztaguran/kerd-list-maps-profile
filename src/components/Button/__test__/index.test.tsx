import React from 'react';
import { render } from '@testing-library/react-native';
import { ButtonMode, ButtonType } from 'enums';
import Button from '../index';

describe('Component: Button', () => {
  it('Should render a default Button', () => {
    const { baseElement } = render(
      <Button
        mode={ButtonMode.Contained}
        type={ButtonType.PlainButton}
        text="Test Button"
        onPress={() => {}}
      />,
    );
    expect(baseElement).toMatchSnapshot();
  });
});
