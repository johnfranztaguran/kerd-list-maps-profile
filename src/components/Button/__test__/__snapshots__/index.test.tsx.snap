// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Button Should render a default Button 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessible={true}
      style={
        Object {
          "backgroundColor": "#38647F",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 2,
          "minWidth": 64,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.75,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 1.5,
        }
      }
      testID="btn-test-button"
    >
      <TouchableHighlight
        accessibilityComponentType="button"
        accessibilityLabel="btn-test-button"
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessibilityTraits="button"
        activeOpacity={0.85}
        delayPressIn={0}
        delayPressOut={100}
        disabled={false}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 4,
            },
          ]
        }
        underlayColor="rgba(255, 255, 255, 0.16)"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  false,
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Array [
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "fontFamily": "OpenSans",
                      "fontSize": 14,
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                      "letterSpacing": 0.25,
                      "lineHeight": 19,
                      "marginBottom": 12,
                      "marginTop": 12,
                      "textAlign": "center",
                      "textTransform": "capitalize",
                    },
                    Object {
                      "color": "#FFFFFF",
                    },
                  ],
                ],
              ]
            }
          >
            Test Button
          </Text>
        </View>
      </TouchableHighlight>
    </View>
  </View>
</View>
`;
